# Project: Wrangling and Analyze Data
#### Arnd-Olav Feindt
---

## Table of contents <a class="anchor" id="Tableofcontents"></a>

- [Data Gathering](#DataGathering)
- [Assessing Data](#AssessingData)
  - [Table overview](#Tableoverview)
  - [All found issues](#Allfoundissues)
  - [Quality issues](#Qualityissues)
  - [Tidiness issues](#Tidinessissues)
- [Cleaning Data](#CleaningData)
- [Storing Data](#StoringData)
- [Analyzing and Visualizing Data](#AnalyzingandVisualizingData)
  - [Visualization](#Visualization)
  - [Insights:](#Insights)
  
---

## Data Gathering <a class="anchor" id="DataGathering"></a>
In the cell below, gather **all** three pieces of data for this project and load them in the notebook. **Note:** the methods required to gather each data are different.
1. Directly download the WeRateDogs Twitter archive data (twitter_archive_enhanced.csv)

In [1]:
# Load libraries
import pandas as pd
import numpy as np
import os
import requests
import tweepy
import json
from timeit import default_timer as timer
from pandas.api.types import is_numeric_dtype
from pandas.api.types import is_object_dtype
from IPython.display import display, Markdown  # To write (have) code and markdown in the same cell


In [2]:
# Load csv into dataframe
wrd_archive = pd.read_csv('twitter-archive-enhanced.csv')

# View head of dataframe
wrd_archive.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

2. Use the Requests library to download the tweet image prediction (image_predictions.tsv)

In [3]:
# Fetch tsv file from Udactiy server
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response = requests.get(url)
with open(url.split('/')[-1], mode='wb') as file:
    file.write(response.content)
    
# Load downloaded tsv file into dataframe
wrd_image_prediction = pd.read_csv('image-predictions.tsv', sep = '\t')

# View head of dataframe
wrd_image_prediction.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

3. Use the Tweepy library to query additional data via the Twitter API (tweet_json.txt)

In [4]:
# Fetch the data from Twitter if you haven't done it already... because it can take 20 - 30 minutes
file_name = 'tweet_json.txt'
if not os.path.isfile(file_name):
    
    consumer_key = 'jK#######################JQ'
    consumer_secret = 'QZ#######################MX'
    access_token = '62#######################oJ'
    access_secret = 'Gr#######################nZ'

    auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
    auth.set_access_token(access_token, access_secret)

    api = tweepy.API(auth, wait_on_rate_limit = True, wait_on_rate_limit_notify = True)

    # Getting tweet details for a single tweet_id from the twitter archive
    tweet = api.get_status(	892420643555336193, tweet_mode = 'extended' )

    # Viewing the json data
    tweet._json

    # Verifying the presence of retweet count of tweet_id in json data
    tweet._json['retweet_count']

    # Verifying the presence of favorite count of tweet_id in json data
    tweet._json['favorite_count']

    # Creating file
    open(file_name, 'w').close()

    # Query Twitter's API for JSON data for each tweet ID in the Twitter archive & writing twitter json data to tweet_json.txt file
    tweet_id_errors = {}
    count = 0
    start = timer()
    with open(file_name, mode = 'w') as file:
        for tweet_id in wrd_archive.tweet_id:
            try:
                tweet = api.get_status(tweet_id, tweet_mode='extended')
                file.write(json.dumps(tweet._json) + '\n')
                print(str(count) + '_' + str(tweet_id))
                count +=1
            except Exception as e:
                print(str(count) + '_' + str(tweet_id) + str(e))
                tweet_id_errors[str(count) + '_' + str(tweet_id)] = tweet._json
                count +=1
    end = timer()
    print('Elapsed Time:' + str(end-start))

In [5]:
# Reading twitter json data from tweet_json.txt file by converting each json string into python dictionary
# and appending them to a list (row by row) and this list of dictionaries will eventually be converted to a pandas DataFrame 

df_list = []
with open('tweet_json.txt', mode = 'r') as file:
    for row in file.readlines():
        tweet_dict = json.loads(row)
        retweet_count = tweet_dict['retweet_count']
        favorite_count = tweet_dict['favorite_count']
        tweet_id = tweet_dict['id']
        df_list.append({'tweet_id': tweet_id, 
                        'retweet_count': retweet_count, 
                        'favorite_count': favorite_count})  

# Creating DataFrame from list of dictionaries
wrd_archive_add = pd.DataFrame(df_list, columns = ['tweet_id', 'retweet_count', 'favorite_count'])

# Top five rows of dataframe
wrd_archive_add.head()

tweet_id  retweet_count  favorite_count
0  892420643555336193           6975           33711
1  892177421306343426           5276           29231
2  891815181378084864           3465           21982
3  891689557279858688           7196           36804
4  891327558926688256           7721           35192

[top](#Tableofcontents)

## Assessing Data <a class="anchor" id="AssessingData"></a>
In this section, detect and document at least **eight (8) quality issues and two (2) tidiness issue**. You must use **both** visual assessment
programmatic assessement to assess the data.

**Note:** pay attention to the following key points when you access the data.

* You only want original ratings (no retweets) that have images. Though there are 5000+ tweets in the dataset, not all are dog ratings and some are retweets.
* Assessing and cleaning the entire dataset completely would require a lot of time, and is not necessary to practice and demonstrate your skills in data wrangling. Therefore, the requirements of this project are only to assess and clean at least 8 quality issues and at least 2 tidiness issues in this dataset.
* The fact that the rating numerators are greater than the denominators does not need to be cleaned. This [unique rating system](http://knowyourmeme.com/memes/theyre-good-dogs-brent) is a big part of the popularity of WeRateDogs.
* You do not need to gather the tweets beyond August 1st, 2017. You can, but note that you won't be able to gather the image predictions for these tweets since you don't have access to the algorithm used.



In [6]:
def visual(dataframe, string_name):
    display(Markdown("##### A. Head from " + string_name))
    print(dataframe.head())
    display(Markdown("##### B. Tail from " + string_name))
    print(dataframe.tail())
    display(Markdown("##### C. Sample of 25 rows from " + string_name))
    df_elements = dataframe.sample(n=25)
    print(df_elements.head(25))

In [18]:
def programmatic(dataframe, string_name):
    display(Markdown(f"##### D. Shape from {string_name}:"))
    print('')
    print(dataframe.shape)
    display(Markdown(f"##### E. Info from {string_name}:"))
    print('')

    print(dataframe.info())
    display(Markdown(f"##### F. isnull sum from {string_name}:"))
    print('')
    print(dataframe.isnull().sum())
    display(Markdown(f"##### G. Duplicates from {string_name}:"))
    print('')
    print(sum(dataframe.duplicated()))
    display(Markdown(f"##### H. Describe from {string_name}:"))
    print('')
    print(dataframe.describe())
    display(Markdown(f"##### I. Nunique from {string_name}:"))
    print('')
    print(dataframe.nunique())

    display(Markdown(f"##### J. Duplicates per column from {string_name}:"))
    for i, col in zip(dataframe.dtypes, dataframe):
        print('')
        print(
            f"Sum of duplicates in {col} = "
            + str(sum(dataframe[col].duplicated()))
            + f"({i})"
        )

    display(Markdown(f"##### K. Sorted numeric values per column in {string_name}:"))
    for i, col in zip(dataframe.dtypes, dataframe):
        print('')
        print(f"Sorted numeric values per column in {col}")

        if is_numeric_dtype(
            dataframe[col]
        ):
            print(np.sort(dataframe[col].unique()))
        else:
            print(col + " is not numeric")

    display(Markdown(f"##### L. Check which column is object type from {string_name}:"))
    for i, col in zip(dataframe.dtypes, dataframe):
        print('')
        if is_object_dtype(dataframe[col]):
            print(str(col), 'is', str(i))
        else:
            print(str(col), 'is not')

    display(Markdown(f"##### M. Count different values per column from {string_name}:"))
    
    for i, col in zip(dataframe.dtypes, dataframe):
        print('')
        print(str(col))
        print(dataframe[col].value_counts())
        print('')


In [19]:
dfs = {'wrd_archive' : wrd_archive,
       'wrd_image_prediction' : wrd_image_prediction,
       'wrd_archive_add':wrd_archive_add
      }
print('__________________________')

for k, v in dfs.items():
    display(Markdown(f'### {k} - Visual assessment'))
    visual(v, k)
    display(Markdown(f'### {k} - Programmatic assessment'))
    programmatic(v, k)
    print ('')

__________________________


### wrd_archive - Visual assessment

##### A. Head from wrd_archive

             tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  r

##### B. Tail from wrd_archive

                tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
2351  666049248165822465                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

 

##### C. Sample of 25 rows from wrd_archive

                tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
517   810896069567610880                    NaN                  NaN   
113   870726314365509632           8.707262e+17           16487760.0   
1611  685325112850124800                    NaN                  NaN   
1232  713175907180089344                    NaN                  NaN   
634   793614319594401792                    NaN                  NaN   
1409  699060279947165696                    NaN                  NaN   
1063  741067306818797568                    NaN                  NaN   
1112  733460102733135873                    NaN                  NaN   
694   786729988674449408                    NaN                  NaN   
316   834931633769889797                    NaN                  NaN   
2184  668988183816871936                    NaN                  NaN   
1263  710117014656950272                    NaN                  NaN   
2071  671115716440031232                    NaN                 

### wrd_archive - Programmatic assessment

##### D. Shape from wrd_archive:


(2356, 17)


##### E. Info from wrd_archive:


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356

##### F. isnull sum from wrd_archive:


tweet_id                         0
in_reply_to_status_id         2278
in_reply_to_user_id           2278
timestamp                        0
source                           0
text                             0
retweeted_status_id           2175
retweeted_status_user_id      2175
retweeted_status_timestamp    2175
expanded_urls                   59
rating_numerator                 0
rating_denominator               0
name                             0
doggo                            0
floofer                          0
pupper                           0
puppo                            0
dtype: int64


##### G. Duplicates from wrd_archive:


0


##### H. Describe from wrd_archive:


           tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+1

##### I. Nunique from wrd_archive:


tweet_id                      2356
in_reply_to_status_id           77
in_reply_to_user_id             31
timestamp                     2356
source                           4
text                          2356
retweeted_status_id            181
retweeted_status_user_id        25
retweeted_status_timestamp     181
expanded_urls                 2218
rating_numerator                40
rating_denominator              18
name                           957
doggo                            2
floofer                          2
pupper                           2
puppo                            2
dtype: int64


##### J. Duplicates per column from wrd_archive:


Sum of duplicates in tweet_id = 0(int64)

Sum of duplicates in in_reply_to_status_id = 2278(float64)

Sum of duplicates in in_reply_to_user_id = 2324(float64)

Sum of duplicates in timestamp = 0(object)

Sum of duplicates in source = 2352(object)

Sum of duplicates in text = 0(object)

Sum of duplicates in retweeted_status_id = 2174(float64)

Sum of duplicates in retweeted_status_user_id = 2330(float64)

Sum of duplicates in retweeted_status_timestamp = 2174(object)

Sum of duplicates in expanded_urls = 137(object)

Sum of duplicates in rating_numerator = 2316(int64)

Sum of duplicates in rating_denominator = 2338(int64)

Sum of duplicates in name = 1399(object)

Sum of duplicates in doggo = 2354(object)

Sum of duplicates in floofer = 2354(object)

Sum of duplicates in pupper = 2354(object)

Sum of duplicates in puppo = 2354(object)


##### K. Sorted numeric values per column in wrd_archive:


Sorted numeric values per column in tweet_id
[666020888022790149 666029285002620928 666033412701032449 ...
 891815181378084864 892177421306343426 892420643555336193]

Sorted numeric values per column in in_reply_to_status_id
[6.65814697e+17 6.67065536e+17 6.67152164e+17 6.67806455e+17
 6.68920717e+17 6.69354383e+17 6.70668383e+17 6.71544874e+17
 6.71561002e+17 6.71729907e+17 6.73715862e+17 6.74468881e+17
 6.74739953e+17 6.74752233e+17 6.74793399e+17 6.74999808e+17
 6.75349384e+17 6.75497103e+17 6.75707330e+17 6.75845657e+17
 6.76588346e+17 6.78021116e+17 6.81339449e+17 6.82788442e+17
 6.84222868e+17 6.84481075e+17 6.84959799e+17 6.85547936e+17
 6.86034025e+17 6.90341254e+17 6.91416866e+17 6.92041935e+17
 6.92417313e+17 6.93572216e+17 6.93642232e+17 6.96488711e+17
 7.03041950e+17 7.03255936e+17 7.04485745e+17 7.07980066e+17
 7.29113531e+17 7.33109485e+17 7.38411920e+17 7.46885919e+17
 7.47648654e+17 7.50180499e+17 7.59099524e+17 7.63865175e+17
 7.66711819e+17 7.72743019e+17 7.97123751e

##### L. Check which column is object type from wrd_archive:


tweet_id is not

in_reply_to_status_id is not

in_reply_to_user_id is not

timestamp is object

source is object

text is object

retweeted_status_id is not

retweeted_status_user_id is not

retweeted_status_timestamp is object

expanded_urls is object

rating_numerator is not

rating_denominator is not

name is object

doggo is object

floofer is object

pupper is object

puppo is object


##### M. Count different values per column from wrd_archive:


tweet_id
892420643555336193    1
687102708889812993    1
687826841265172480    1
687818504314159109    1
687807801670897665    1
                     ..
775085132600442880    1
774757898236878852    1
774639387460112384    1
774314403806253056    1
666020888022790149    1
Name: tweet_id, Length: 2356, dtype: int64


in_reply_to_status_id
6.671522e+17    2
8.862664e+17    1
6.920419e+17    1
6.827884e+17    1
6.842229e+17    1
               ..
8.116272e+17    1
8.131273e+17    1
8.211526e+17    1
8.233264e+17    1
6.670655e+17    1
Name: in_reply_to_status_id, Length: 77, dtype: int64


in_reply_to_user_id
4.196984e+09    47
2.195506e+07     2
2.281182e+09     1
1.132119e+08     1
1.637468e+07     1
4.670367e+08     1
1.198989e+09     1
2.878549e+07     1
2.319108e+09     1
3.589728e+08     1
4.717297e+09     1
1.584641e+07     1
7.305050e+17     1
2.916630e+07     1
2.918590e+08     1
1.185634e+07     1
2.068372e+07     1
1.582854e+09     1
4.738443e+07     1
3.058208e+07     1
2.625

### wrd_image_prediction - Visual assessment

##### A. Head from wrd_image_prediction

             tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheep

##### B. Tail from wrd_image_prediction

                tweet_id                                          jpg_url  \
2070  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
2071  891689557279858688  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg   
2072  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
2073  892177421306343426  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
2074  892420643555336193  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

      img_num           p1   p1_conf  p1_dog                  p2   p2_conf  \
2070        2       basset  0.555712    True    English_springer  0.225770   
2071        1  paper_towel  0.170278   False  Labrador_retriever  0.168086   
2072        1    Chihuahua  0.716012    True            malamute  0.078253   
2073        1    Chihuahua  0.323581    True            Pekinese  0.090647   
2074        1       orange  0.097049   False               bagel  0.085851   

      p2_dog                           p3   p3_conf  p3_dog  
2070  

##### C. Sample of 25 rows from wrd_image_prediction

                tweet_id                                            jpg_url  \
374   672995267319328768    https://pbs.twimg.com/media/CVb1mRiWcAADBsE.jpg   
36    666447344410484738    https://pbs.twimg.com/media/CT-yU5QWwAEjLX5.jpg   
524   676603393314578432    https://pbs.twimg.com/media/CWPHMqKVAAAE78E.jpg   
477   675166823650848770    https://pbs.twimg.com/media/CV6spB7XAAIpMyP.jpg   
1153  732726085725589504    https://pbs.twimg.com/media/CisqdVcXEAE3iW7.jpg   
1118  725842289046749185    https://pbs.twimg.com/media/ChK1tdBWwAQ1flD.jpg   
1246  747594051852075008    https://pbs.twimg.com/media/Cl_80k5WkAEbo9m.jpg   
1338  758467244762497024  https://pbs.twimg.com/ext_tw_video_thumb/75846...   
2034  883838122936631299    https://pbs.twimg.com/media/DEQGFgAXUAAEvfi.jpg   
1425  772193107915964416    https://pbs.twimg.com/media/Crdhh_1XEAAHKHi.jpg   
1757  825147591692263424    https://pbs.twimg.com/media/C3ODWpfXAAAP1fb.jpg   
1546  792394556390137856    https://pbs.twimg.com/me

### wrd_image_prediction - Programmatic assessment

##### D. Shape from wrd_image_prediction:


(2075, 12)


##### E. Info from wrd_image_prediction:


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB
None


##### F. isnull sum from wrd_image_prediction:


tweet_id    0
jpg_url     0
img_num     0
p1          0
p1_conf     0
p1_dog      0
p2          0
p2_conf     0
p2_dog      0
p3          0
p3_conf     0
p3_dog      0
dtype: int64


##### G. Duplicates from wrd_image_prediction:


0


##### H. Describe from wrd_image_prediction:


           tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01


##### I. Nunique from wrd_image_prediction:


tweet_id    2075
jpg_url     2009
img_num        4
p1           378
p1_conf     2006
p1_dog         2
p2           405
p2_conf     2004
p2_dog         2
p3           408
p3_conf     2006
p3_dog         2
dtype: int64


##### J. Duplicates per column from wrd_image_prediction:


Sum of duplicates in tweet_id = 0(int64)

Sum of duplicates in jpg_url = 66(object)

Sum of duplicates in img_num = 2071(int64)

Sum of duplicates in p1 = 1697(object)

Sum of duplicates in p1_conf = 69(float64)

Sum of duplicates in p1_dog = 2073(bool)

Sum of duplicates in p2 = 1670(object)

Sum of duplicates in p2_conf = 71(float64)

Sum of duplicates in p2_dog = 2073(bool)

Sum of duplicates in p3 = 1667(object)

Sum of duplicates in p3_conf = 69(float64)

Sum of duplicates in p3_dog = 2073(bool)


##### K. Sorted numeric values per column in wrd_image_prediction:


Sorted numeric values per column in tweet_id
[666020888022790149 666029285002620928 666033412701032449 ...
 891815181378084864 892177421306343426 892420643555336193]

Sorted numeric values per column in jpg_url
jpg_url is not numeric

Sorted numeric values per column in img_num
[1 2 3 4]

Sorted numeric values per column in p1
p1 is not numeric

Sorted numeric values per column in p1_conf
[0.0443334 0.0553794 0.0590326 ... 0.999962  0.999984  1.       ]

Sorted numeric values per column in p1_dog
[False  True]

Sorted numeric values per column in p2
p2 is not numeric

Sorted numeric values per column in p2_conf
[1.01130e-08 1.00288e-05 1.44895e-05 ... 4.64816e-01 4.67678e-01
 4.88014e-01]

Sorted numeric values per column in p2_dog
[False  True]

Sorted numeric values per column in p3
p3 is not numeric

Sorted numeric values per column in p3_conf
[1.74017e-10 2.16090e-07 5.59504e-07 ... 2.70673e-01 2.71042e-01
 2.73419e-01]

Sorted numeric values per column in p3_dog
[False  True]


##### L. Check which column is object type from wrd_image_prediction:


tweet_id is not

jpg_url is object

img_num is not

p1 is object

p1_conf is not

p1_dog is not

p2 is object

p2_conf is not

p2_dog is not

p3 is object

p3_conf is not

p3_dog is not


##### M. Count different values per column from wrd_image_prediction:


tweet_id
666020888022790149    1
761292947749015552    1
767122157629476866    1
766793450729734144    1
766693177336135680    1
                     ..
684097758874210310    1
683857920510050305    1
683852578183077888    1
683849932751646720    1
892420643555336193    1
Name: tweet_id, Length: 2075, dtype: int64


jpg_url
https://pbs.twimg.com/media/CZhn-QAWwAASQan.jpg                                            2
https://pbs.twimg.com/media/Cq9guJ5WgAADfpF.jpg                                            2
https://pbs.twimg.com/ext_tw_video_thumb/807106774843039744/pu/img/8XZg1xW35Xp2J6JW.jpg    2
https://pbs.twimg.com/media/CU1zsMSUAAAS0qW.jpg                                            2
https://pbs.twimg.com/media/CsrjryzWgAAZY00.jpg                                            2
                                                                                          ..
https://pbs.twimg.com/media/CXrmMSpUwAAdeRj.jpg                                            1
https://pbs.twimg.com/

### wrd_archive_add - Visual assessment

##### A. Head from wrd_archive_add

             tweet_id  retweet_count  favorite_count
0  892420643555336193           6975           33711
1  892177421306343426           5276           29231
2  891815181378084864           3465           21982
3  891689557279858688           7196           36804
4  891327558926688256           7721           35192


##### B. Tail from wrd_archive_add

                tweet_id  retweet_count  favorite_count
2322  666049248165822465             36              88
2323  666044226329800704            115             246
2324  666033412701032449             36             100
2325  666029285002620928             39             112
2326  666020888022790149            420            2285


##### C. Sample of 25 rows from wrd_archive_add

                tweet_id  retweet_count  favorite_count
1603  684481074559381504           1069            3559
2263  667160273090932737             50             219
1412  696886256886657024           1626            4441
1022  743210557239623680           1232            3551
2310  666102155909144576             11              66
492   811985624773361665           1304            6893
2197  668274247790391296            196             718
2222  667806454573760512            416             926
266   839549326359670784           7057           25521
978   747648653817413632           5305           11932
371   826476773533745153           3885           17330
1406  697270446429966336           1712            4340
737   778408200802557953           3994           12830
1541  687807801670897665            657            2197
1448  693629975228977152            719            2257
250   842163532590374912           5249           22687
915   752701944171524096           2609         

### wrd_archive_add - Programmatic assessment

##### D. Shape from wrd_archive_add:


(2327, 3)


##### E. Info from wrd_archive_add:


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2327 entries, 0 to 2326
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   tweet_id        2327 non-null   int64
 1   retweet_count   2327 non-null   int64
 2   favorite_count  2327 non-null   int64
dtypes: int64(3)
memory usage: 54.7 KB
None


##### F. isnull sum from wrd_archive_add:


tweet_id          0
retweet_count     0
favorite_count    0
dtype: int64


##### G. Duplicates from wrd_archive_add:


0


##### H. Describe from wrd_archive_add:


           tweet_id  retweet_count  favorite_count
count  2.327000e+03    2327.000000     2327.000000
mean   7.417930e+17    2459.823378     7029.201117
std    6.820795e+16    4165.838143    10923.364445
min    6.660209e+17       1.000000        0.000000
25%    6.781394e+17     492.500000     1220.500000
50%    7.178418e+17    1144.000000     3040.000000
75%    7.986547e+17    2845.000000     8568.000000
max    8.924206e+17   70366.000000   144304.000000


##### I. Nunique from wrd_archive_add:


tweet_id          2327
retweet_count     1633
favorite_count    1971
dtype: int64


##### J. Duplicates per column from wrd_archive_add:


Sum of duplicates in tweet_id = 0(int64)

Sum of duplicates in retweet_count = 694(int64)

Sum of duplicates in favorite_count = 356(int64)


##### K. Sorted numeric values per column in wrd_archive_add:


Sorted numeric values per column in tweet_id
[666020888022790149 666029285002620928 666033412701032449 ...
 891815181378084864 892177421306343426 892420643555336193]

Sorted numeric values per column in retweet_count
[    1     2     3 ... 51452 52659 70366]

Sorted numeric values per column in favorite_count
[     0     45     59 ... 111241 123716 144304]


##### L. Check which column is object type from wrd_archive_add:


tweet_id is not

retweet_count is not

favorite_count is not


##### M. Count different values per column from wrd_archive_add:


tweet_id
892420643555336193    1
688828561667567616    1
687460506001633280    1
687399393394311168    1
687317306314240000    1
                     ..
772877495989305348    1
772826264096874500    1
772615324260794368    1
772581559778025472    1
666020888022790149    1
Name: tweet_id, Length: 2327, dtype: int64


retweet_count
819     5
521     5
406     5
50      5
669     5
       ..
3614    1
2707    1
4261    1
2328    1
420     1
Name: retweet_count, Length: 1633, dtype: int64


favorite_count
0        160
1148       4
371        4
1275       3
657        3
        ... 
4724       1
3919       1
13356      1
5070       1
2285       1
Name: favorite_count, Length: 1971, dtype: int64




In [13]:
num

NameError: name 'num' is not defined

[top](#Tableofcontents)

### Table overview <a class="anchor" id="Tableoverview"></a>
 - **wrd_archive - contains twitter archive data of WeRateDogs**
 - **wrd_image_prediction - contains twitter api data (retweets and likes) of WeRateDogs**
 - **wrd_archive_add - contains tweet image prediction data**
---
#### What should the type and possible values of the columns be
##### wrd_archive
|Column|actual type|target type | possible values|
|----------------------------|-------|----------|-------|
| tweet_id                   |  int64 | string | 1 != empty  |
| in_reply_to_status_id      |  float64 | int64 |   |
| in_reply_to_user_id        |  float64 | int64 |   |
| timestamp                  |  object | Timestamp |   |
| source                     |  object | string |   |
| text                       |  object | string |   |
| retweeted_status_id        | float64  | int64 |   |
| retweeted_status_user_id   | float64  | int64 |   |
| retweeted_status_timestamp | object  | int64 |   |
| expanded_urls              |  object | string |   |
| rating_numerator           | int64  | int64 |   |
| rating_denominator         |  int64 | int64 |   |
| name                       | object  | string |   |
| doggo                      |  object | string |   |
| floofer                    | object  | string |   |
| pupper                     | object  | string |   |
| puppo                      | object  | string |    |

##### wrd_image_prediction
|Column|actual type|target type | possible values|
|----------|-----------------------|---|---| 
| tweet_id  |int64 | string  |   |
| jpg_url  |object | string  |   |
| img_num  | int64| int64   |   |
| p1       |object | string  |   |
| p1_conf  | float64| float64 |   |
| p1_dog   |bool | bool    |   |
| p2       | object| string  |   |
| p2_conf  | float64| float64 |   |
| p2_dog   | bool| bool    |   |
| p3       |object | string  |   |
| p3_conf  |float64 | float64 |   |
| p3_dog   | bool| bool    |   |

##### wrd_archive_add
|Column|actual type|target type | possible values|
|----------|-----------------------|---|---| 
| tweet_id  | int64 | string  |   |
| retweet_count  | int64 | int64   |   |
| favorite_count | int64 | int64  |   |

### All found issues <a class="anchor" id="Allfoundissues"></a>

| **Issue ID** | **Issue Type** | **Method** | **Dimension** | **Table** | **Column** | **Description** |
|---|---|---|---|---|---|---|
| 1 | Q | P | Completeness | wrd_archive | in_reply_to_status_id<br>in_reply_to_user_id | Observations for Uninterested reply columns present(78). |
| 2 | Q | P | Completeness | wrd_archive | retweeted_status_timestamp<br>retweeted_status_id<br>retweeted_status_user_id | Observations for Uninterested retweet columns present(181). |
| 3 | Q | V | Completeness | wrd_archive | in_reply_to_status_id<br>in_reply_to_user_id<br>retweeted_status_timestamp<br>retweeted_status_id<br>retweeted_status_user_id | Unnecessary columns |
| 4 | Q | P | Validity | wrd_archive<br>wrd_image_prediction<br>wrd_archive_add | tweet_id | Type *int64* instead of *string* |
| 5 | Q | P | Validity | wrd_archive | timestamp<br>retweeted_status_timestamp | Type *object* instead of *timestamp* |
| 6 | Q | V | Accuracy | wrd_archive | rating_numerator | Inaccurate values |
| 7 | Q | V | Accuracy | wrd_archive | rating_denominator | Inaccurate values != 10 |
| 8 | Q | V | Consistency | wrd_archive | rating_numerator<br>rating_denominator | Inconsistent rating |
| 9 | Q | P | Consistency | wrd_archive | expanded_urls | Sometimes more than one value |
| 10 | Q | V | Validity | wrd_archive | name | Invalid names or non-names |
| 11 | Q | P | Consistency | wrd_archive | text | ‘&’ and ‘\n’ in (sub)strings |
| 12 | Q | V | Consistency | wrd_archive | doggo<br>floofer<br>pupper<br>puppo | Categorical variables represented as separate column. |
| 13 | Q | V | Validity | wrd_image_prediction | p1_conf, p2_conf, p3_conf | Drop the rows with FALSE in all these columns |
| 14 | Q | V | Consistency | wrd_image_prediction | p1<br>p2<br>p3 | One Prediction data with TRUE and higher confidence is enough among the top three predictions. |
| 15 | T | P | - | wrd_archive | source | Multiple data values in one column (HTML tags URL and content) |
| 16 | T | P | - | wrd_archive | text | Multiple data values in one column (Text and URL) |
| 17 | T | P | - | wrd_archive<br>wrd_image_prediction<br>wrd_archive_add | tweet_id | Duplicate column in three DataFrame |
| 18 | T | V | - | wrd_image_prediction | p1, p1_dog, p1_conf<br>p2, p2_dog, p2_conf<br>p3, p3_dog, p3_conf | For quality I will create two new columns for image prediction and confidence level. After I will drop p1, p2, p3, p1_conf, p2_conf, p3_conf, p1_dog, p2_conf and p3_conf |
| 19 | Q | V | Consistency | wrd_image_prediction | p1<br>p2<br>p3 | Replace '_' with space in breed and Replace '-' with space/no space for breeds wherever needed based on actual name browsing |
| 20 | T | V | - | wrd_image_prediction | breed | Capitalize the breed strings |

### Quality issues <a class="anchor" id="Qualityissues"></a>

I choose the issues 1,2,3,4,5,6,7 and 12

[top](#Tableofcontents)

### Tidiness issues <a class="anchor" id="Tidinessissues"></a>

I choose the issues 15, 16 and 18

[top](#Tableofcontents)

## Cleaning Data <a class="anchor" id="CleaningData"></a>
In this section, clean **all** of the issues you documented while assessing. 

**Note:** Make a copy of the original data before cleaning. Cleaning includes merging individual pieces of data according to the rules of [tidy data](https://cran.r-project.org/web/packages/tidyr/vignettes/tidy-data.html). The result should be a high-quality and tidy master pandas DataFrame (or DataFrames, if appropriate).

[top](#Tableofcontents)

In [32]:
# Make copies of original pieces of data
wrd_archive_clean = wrd_archive.copy()
wrd_image_prediction_clean = wrd_image_prediction.copy()
wrd_archive_add_clean = wrd_archive_add.copy()

### Issue 1-Q

#### Define
**wrd_archive_clean/in_reply_to_status_id, in_reply_to_user_id**	Not necessary observations of uninterested reply columns present. I will drop the rows with input in these columns.

#### Code

In [33]:
# After counting the affected rows I will delete them within the 'Test' block bellow
print(len(wrd_archive_clean.query('in_reply_to_status_id != "NaN"')))
print(len(wrd_archive_clean.query('in_reply_to_user_id != "NaN"')))

78
78


#### Test

In [34]:
print(wrd_archive_clean.shape)
wrd_archive_clean = wrd_archive_clean[~wrd_archive_clean['in_reply_to_status_id'].notna()]
print(wrd_archive_clean.shape)

(2356, 17)
(2278, 17)


### Issue 2-Q

#### Define
**wrd_archive_clean/retweeted_status_timestamp, retweeted_status_id, retweeted_status_user_id**

Not necessary observations of uninterested retweet columns present.
I will drop the rows with input in these columns.
I am facing a problem here with counting the NaN in column 'retweeted_status_timestamp', so I let it.

#### Code

In [35]:
# After counting the affected rows I will delete them within the 'Test' block bellow
print(len(wrd_archive_clean[~wrd_archive_clean['retweeted_status_id'].isnull()]))
print(len(wrd_archive_clean[~wrd_archive_clean['retweeted_status_user_id'].isnull()]))


181
181


#### Test

In [36]:
print(wrd_archive_clean.shape)
wrd_archive_clean = wrd_archive_clean[wrd_archive_clean['retweeted_status_id'].isnull()]
wrd_archive_clean = wrd_archive_clean[wrd_archive_clean['retweeted_status_user_id'].isnull()]
print(wrd_archive_clean.shape)

(2278, 17)
(2097, 17)


### Issue 3-Q

#### Define
**wrd_archive_clean/in_reply_to_status_id, in_reply_to_user_id, retweeted_status_timestamp, retweeted_status_id, retweeted_status_user_id**

Unnecessary columns.

I will drop these columns

#### Code

In [37]:
print(wrd_archive_clean.shape)
wrd_archive_clean.drop(['in_reply_to_status_id',
                        'in_reply_to_user_id',
                        'retweeted_status_timestamp',
                        'retweeted_status_id',
                        'retweeted_status_user_id'
                        ], axis=1, inplace=True)

(2097, 17)


#### Test

In [38]:
print(wrd_archive_clean.shape, "\n")
print(wrd_archive_clean.info())

(2097, 12) 

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2097 entries, 0 to 2355
Data columns (total 12 columns):
tweet_id              2097 non-null int64
timestamp             2097 non-null object
source                2097 non-null object
text                  2097 non-null object
expanded_urls         2094 non-null object
rating_numerator      2097 non-null int64
rating_denominator    2097 non-null int64
name                  2097 non-null object
doggo                 2097 non-null object
floofer               2097 non-null object
pupper                2097 non-null object
puppo                 2097 non-null object
dtypes: int64(3), object(9)
memory usage: 213.0+ KB
None


### Issue 4-Q

#### Define
**wrd_archive, wrd_image_prediction, wrd_archive_add/tweet_id**	Type int64 instead of string
I will change the type of the columns to string.

#### Code

In [39]:
wrd_archive_clean.tweet_id = wrd_archive_clean.tweet_id.astype(str)
wrd_image_prediction_clean.tweet_id = wrd_image_prediction_clean.tweet_id.astype(str)
wrd_archive_add_clean.tweet_id = wrd_archive_add_clean.tweet_id.astype(str)

#### Test

In [40]:
print(wrd_archive_clean.tweet_id.apply(type), "\n")
print(wrd_image_prediction_clean.tweet_id.apply(type), "\n")
print(wrd_archive_add_clean.tweet_id.apply(type), "\n")

0       <class 'str'>
1       <class 'str'>
2       <class 'str'>
3       <class 'str'>
4       <class 'str'>
5       <class 'str'>
6       <class 'str'>
7       <class 'str'>
8       <class 'str'>
9       <class 'str'>
10      <class 'str'>
11      <class 'str'>
12      <class 'str'>
13      <class 'str'>
14      <class 'str'>
15      <class 'str'>
16      <class 'str'>
17      <class 'str'>
18      <class 'str'>
20      <class 'str'>
21      <class 'str'>
22      <class 'str'>
23      <class 'str'>
24      <class 'str'>
25      <class 'str'>
26      <class 'str'>
27      <class 'str'>
28      <class 'str'>
29      <class 'str'>
31      <class 'str'>
            ...      
2326    <class 'str'>
2327    <class 'str'>
2328    <class 'str'>
2329    <class 'str'>
2330    <class 'str'>
2331    <class 'str'>
2332    <class 'str'>
2333    <class 'str'>
2334    <class 'str'>
2335    <class 'str'>
2336    <class 'str'>
2337    <class 'str'>
2338    <class 'str'>
2339    <class 'str'>
2340    <c

### Issue 5-Q

#### Define
**wrd_archive_clean/timestamp, retweeted_status_timestamp (already dropped)** Type *object* instead of *timestamp*
I will change the column type to `timestamp`

#### Code

In [41]:
wrd_archive_clean.timestamp = pd.to_datetime(wrd_archive_clean.timestamp)

#### Test

In [42]:
print(wrd_archive_clean.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2097 entries, 0 to 2355
Data columns (total 12 columns):
tweet_id              2097 non-null object
timestamp             2097 non-null datetime64[ns]
source                2097 non-null object
text                  2097 non-null object
expanded_urls         2094 non-null object
rating_numerator      2097 non-null int64
rating_denominator    2097 non-null int64
name                  2097 non-null object
doggo                 2097 non-null object
floofer               2097 non-null object
pupper                2097 non-null object
puppo                 2097 non-null object
dtypes: datetime64[ns](1), int64(2), object(9)
memory usage: 213.0+ KB
None


### Issue 6-Q

#### Define
**wrd_archive_clean/rating_denominator**
Inaccurate values != 10. As these are just 17 rows I will drop them.

#### Code

In [43]:
# After counting the affected rows I will delete them within the 'Test' block bellow
len(wrd_archive_clean.query('rating_denominator != 10'))

17

#### Test

In [44]:
print(wrd_archive_clean.shape)
wrd_archive_clean = wrd_archive_clean[wrd_archive_clean.rating_denominator == 10]
print(wrd_archive_clean.shape)

(2097, 12)
(2080, 12)


### Issue 7-Q

#### Define
**wrd_archive_clean/rating_numerator**

We have some numerator >= 15 rows, there is a big gap between 14 and 26. This seems unrealisted watching the distribution. As these are just 5 rows I will drop them.

#### Code

In [45]:
wrd_archive_clean['rating_numerator'].value_counts().sort_index(ascending = False)

1776      1
420       1
75        1
27        1
26        1
14       38
13      287
12      486
11      413
10      436
9       152
8        98
7        51
6        32
5        34
4        15
3        19
2         9
1         4
0         1
Name: rating_numerator, dtype: int64

#### Test

In [46]:
# After counting the affected rows I will delete them within the 'Test' block bellow
print(len(wrd_archive_clean.query('rating_numerator >= 15')),"\n")

print(wrd_archive_clean.shape)
wrd_archive_clean = wrd_archive_clean[wrd_archive_clean.rating_numerator <= 15]
print(wrd_archive_clean.shape)

5 

(2080, 12)
(2075, 12)


### Issue 12-Q

#### Define

**wrd_archive_clean/doggo, floofer, pupper, puppo**

Categorical variables represented as separate columns.

I will create a column dog_stage with the values from these columns and drop them afterwards.

#### Code

In [47]:
import re
wrd_archive_clean['dog_stage'] = wrd_archive_clean.apply(lambda x : re.sub('None,*','',(x.doggo +','+ x.floofer + x.puppo + x.pupper)).rstrip(',') 
                     if re.sub('None','',(x.doggo + x.floofer + x.pupper + x.puppo)) != ''
                     else 'None', axis=1)

columns_to_drop = ['doggo',
                   'floofer',
                   'pupper',
                   'puppo'
                  ]
wrd_archive_clean.drop(columns = columns_to_drop, inplace = True)


#### Test

In [48]:
wrd_archive_clean

tweet_id           timestamp  \
0     892420643555336193 2017-08-01 16:23:56   
1     892177421306343426 2017-08-01 00:17:27   
2     891815181378084864 2017-07-31 00:18:03   
3     891689557279858688 2017-07-30 15:58:51   
4     891327558926688256 2017-07-29 16:00:24   
5     891087950875897856 2017-07-29 00:08:17   
6     890971913173991426 2017-07-28 16:27:12   
7     890729181411237888 2017-07-28 00:22:40   
8     890609185150312448 2017-07-27 16:25:51   
9     890240255349198849 2017-07-26 15:59:51   
10    890006608113172480 2017-07-26 00:31:25   
11    889880896479866881 2017-07-25 16:11:53   
12    889665388333682689 2017-07-25 01:55:32   
13    889638837579907072 2017-07-25 00:10:02   
14    889531135344209921 2017-07-24 17:02:04   
15    889278841981685760 2017-07-24 00:19:32   
16    888917238123831296 2017-07-23 00:22:39   
17    888804989199671297 2017-07-22 16:56:37   
18    888554962724278272 2017-07-22 00:23:06   
20    888078434458587136 2017-07-20 16:49:33   
21    887705289381826560 2017-07-19 16:06:48   
22    887517139158093824 2017-07-19 03:39:09   
23    887473957103951883 2017-07-19 00:47:34   
24    887343217045368832 2017-07-18 16:08:03   
25    887101392804085760 2017-07-18 00:07:08   
26    886983233522544640 2017-07-17 16:17:36   
27    886736880519319552 2017-07-16 23:58:41   
28    886680336477933568 2017-07-16 20:14:00   
29    886366144734445568 2017-07-15 23:25:31   
31    886258384151887873 2017-07-15 16:17:19   
...                  ...                 ...   
2325  666418789513326592 2015-11-17 00:53:15   
2326  666411507551481857 2015-11-17 00:24:19   
2327  666407126856765440 2015-11-17 00:06:54   
2328  666396247373291520 2015-11-16 23:23:41   
2329  666373753744588802 2015-11-16 21:54:18   
2330  666362758909284353 2015-11-16 21:10:36   
2331  666353288456101888 2015-11-16 20:32:58   
2332  666345417576210432 2015-11-16 20:01:42   
2333  666337882303524864 2015-11-16 19:31:45   
2334  666293911632134144 2015-11-16 16:37:02   
2336  666273097616637952 2015-11-16 15:14:19   
2337  666268910803644416 2015-11-16 14:57:41   
2338  666104133288665088 2015-11-16 04:02:55   
2339  666102155909144576 2015-11-16 03:55:04   
2340  666099513787052032 2015-11-16 03:44:34   
2341  666094000022159362 2015-11-16 03:22:39   
2342  666082916733198337 2015-11-16 02:38:37   
2343  666073100786774016 2015-11-16 01:59:36   
2344  666071193221509120 2015-11-16 01:52:02   
2345  666063827256086533 2015-11-16 01:22:45   
2346  666058600524156928 2015-11-16 01:01:59   
2347  666057090499244032 2015-11-16 00:55:59   
2348  666055525042405380 2015-11-16 00:49:46   
2349  666051853826850816 2015-11-16 00:35:11   
2350  666050758794694657 2015-11-16 00:30:50   
2351  666049248165822465 2015-11-16 00:24:50   
2352  666044226329800704 2015-11-16 00:04:52   
2353  666033412701032449 2015-11-15 23:21:54   
2354  666029285002620928 2015-11-15 23:05:30   
2355  666020888022790149 2015-11-15 22:32:08   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
5     <a href="http://twitter.com/download/iphone" r...   
6     <a href="http://twitter.com/download/iphone" r...   
7     <a href="http://twitter.com/download/iphone" r...   
8     <a href="http://twitter.com/download/iphone" r...   
9     <a href="http://twitter.com/download/iphone" r...   
10    <a href="http://twitter.com/download/iphone" r...   
11    <a href="http://twitter.com/download/iphone" r...   
12    <a href="http://twitter.com/download/iphone" r...   
13    <a href="http://twitter.com/download/iphone" r...   
14    <a href="http://twitter.com/download/iphone" r...   
15    <a href="http://twitter.com/download/iphone" r...   
16    <a href="http://twitter.com/do

### Issue 7-Q

#### Define

**wrd_image_prediction_clean/p1, p1_dog, p1_conf, p2, p2_dog, p2_conf, p3, p3_dog, p3_conf**

For quality I will create two new columns for image prediction and confidence level. After I will drop p1, p2, p3, p1_conf, p2_conf, p3_conf, p1_dog, p2_conf and p3_conf

#### Define


#### Code

In [51]:
# Create a list to store the best prediction and confidence values
prediction_list = []

# Define a function to perform the extraction process
def extract_breed_info(row):
    """
        Extracts the best prediction and confidence value from passed row.
        Params:
            row: a row from the dataframe of interest.
        Output:
            A dictionary containing prediction and confidence appended into prediction list.
            Prints a status update of extraction process.
    """
    if row.p1_dog:
        prediction_list.append({'breed': row.p1,'confidence': row.p1_conf})
    elif row.p2_dog:
        prediction_list.append({'breed': row.p2,'confidence': row.p2_conf})
    elif row.p3_dog:
        prediction_list.append({'breed': row.p3,'confidence': row.p3_conf})
    else:
        prediction_list.append({'breed': 'Unknown','confidence': 0})
        
    return 'Info extracted to prediction list'

In [53]:
# Run the extraction process
wrd_image_prediction_clean.apply(extract_breed_info, axis=1)

0       Info extracted to prediction list
1       Info extracted to prediction list
2       Info extracted to prediction list
3       Info extracted to prediction list
4       Info extracted to prediction list
5       Info extracted to prediction list
6       Info extracted to prediction list
7       Info extracted to prediction list
8       Info extracted to prediction list
9       Info extracted to prediction list
10      Info extracted to prediction list
11      Info extracted to prediction list
12      Info extracted to prediction list
13      Info extracted to prediction list
14      Info extracted to prediction list
15      Info extracted to prediction list
16      Info extracted to prediction list
17      Info extracted to prediction list
18      Info extracted to prediction list
19      Info extracted to prediction list
20      Info extracted to prediction list
21      Info extracted to prediction list
22      Info extracted to prediction list
23      Info extracted to predicti

In [56]:
# Assign the values in prediction list into new columns in predictions_clean
wrd_image_prediction_clean[['breed', 'confidence']] = pd.DataFrame(prediction_list)

# Round confidence to three decimal places
wrd_image_prediction_clean.confidence = round(wrd_image_prediction_clean.confidence, 3)

In [57]:
# Verify the extraction process
wrd_image_prediction_clean.iloc[:, 3:].sample(5)

p1   p1_conf  p1_dog               p2   p2_conf  p2_dog  \
635    Irish_setter  0.406047    True   cocker_spaniel  0.345646    True   
1748          dingo  0.280949   False  German_shepherd  0.194044    True   
119       Chihuahua  0.664834    True      cowboy_boot  0.060343   False   
1556        doormat  0.874431   False   French_bulldog  0.018759    True   
1760  shopping_cart  0.995941   False  shopping_basket  0.004057   False   

               p3       p3_conf  p3_dog            breed  confidence  
635      Airedale  1.479120e-01    True     Irish_setter       0.406  
1748     Pembroke  1.200510e-01    True  German_shepherd       0.194  
119   giant_panda  5.983750e-02   False        Chihuahua       0.665  
1556  Boston_bull  1.513440e-02    True   French_bulldog       0.019  
1760    mousetrap  8.832830e-07   False          Unknown       0.000

In [87]:
#create new columns
wrd_image_prediction_clean['dog_breed'] = dog_breed
wrd_image_prediction_clean['confidence_level'] = confidence_level

ValueError: Length of values does not match length of index

In [58]:
# Create a list of unwanted columns
unwanted_columns = ['p1','p2', 'p3','p1_conf', 'p2_conf', 'p3_conf','p1_dog', 'p2_dog', 'p3_dog', 'img_num']

# Drop all unwanted columns
wrd_image_prediction_clean.drop(columns=unwanted_columns, inplace=True)

#### Test

In [60]:
wrd_image_prediction_clean.head(5)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

                    breed  confidence  
0  Welsh_springer_spaniel       0.465  
1                 redbone       0.507  
2         German_shepherd       0.596  
3     Rhodesian_ridgeback       0.408  
4      miniature_pinscher       0.560

### Issue 8-Q

#### Define:
**wrd_image_prediction_clean/dog_breed**

Replace '_' with space in breed and Replace '-' with space/no space for breeds wherever needed based on actual name browsing

#### Code

In [61]:
# Remove all underscores and format the breed text to titlecase.
wrd_image_prediction_clean.breed = wrd_image_prediction_clean.breed.str.replace('_', ' ').str.title()

#### Test

In [62]:
wrd_image_prediction_clean.breed.unique()


array(['Welsh Springer Spaniel', 'Redbone', 'German Shepherd',
       'Rhodesian Ridgeback', 'Miniature Pinscher', 'Bernese Mountain Dog',
       'Unknown', 'Chow', 'Golden Retriever', 'Miniature Poodle',
       'Gordon Setter', 'Walker Hound', 'Pug', 'Bloodhound', 'Lhasa',
       'English Setter', 'Italian Greyhound', 'Maltese Dog',
       'Newfoundland', 'Malamute', 'Soft-Coated Wheaten Terrier',
       'Chihuahua', 'Black-And-Tan Coonhound', 'Toy Terrier',
       'Blenheim Spaniel', 'Pembroke', 'Irish Terrier',
       'Chesapeake Bay Retriever', 'Curly-Coated Retriever', 'Dalmatian',
       'Ibizan Hound', 'Border Collie', 'Labrador Retriever',
       'Miniature Schnauzer', 'Airedale', 'Rottweiler',
       'West Highland White Terrier', 'Toy Poodle', 'Giant Schnauzer',
       'Vizsla', 'Siberian Husky', 'Papillon', 'Saint Bernard',
       'Tibetan Terrier', 'Borzoi', 'Beagle', 'Yorkshire Terrier',
       'Pomeranian', 'Kuvasz', 'Flat-Coated Retriever',
       'Norwegian Elkhound', '

### Issue 18-T

#### Define:
**wrd_image_prediction/**

Merge all DataFrames into one Master DataFrame

#### Code

In [63]:
# Merge archive clean and prediction clean into master df
master_df = pd.merge(wrd_archive_clean, wrd_image_prediction, on='tweet_id', how='inner')

# Merge json clean into master df
master_df = pd.merge(master_df, wrd_archive_add, on='tweet_id', how='inner')

# Check results 
master_df.info()

ValueError: You are trying to merge on object and int64 columns. If you wish to proceed you should use pd.concat

#### Test

[top](#Tableofcontents)

## Storing Data <a class="anchor" id="StoringData"></a>
Save gathered, assessed, and cleaned master dataset to a CSV file named "twitter_archive_master.csv".

[top](#Tableofcontents)

## Analyzing and Visualizing Data <a class="anchor" id="AnalyzingandVisualizingData"></a>
In this section, analyze and visualize your wrangled data. You must produce at least **three (3) insights and one (1) visualization.**

[top](#Tableofcontents)

### Insights: <a class="anchor" id="Insights"></a>
1.

2.

3.

[top](#Tableofcontents)

### Visualization <a class="anchor" id="Visualization"></a>

[top](#Tableofcontents)